src/ConfigParser.cpp

```diff
@@ -0,0 +1,213 @@
+  using nlohmann::json;
+
+  if (auto KafkaMaybe = find<json>("kafka", Json)) {
+    auto Kafka = KafkaMaybe.inner();
+    if (auto BrokerMaybe = find<json>("broker", Kafka)) {
+      auto Broker = BrokerMaybe.inner();
+      for (auto Property = Broker.begin(); Property != Broker.end();
+           ++Property) {
+        auto const Key = Property.key();
+        if (Key.find("___") == 0) {
+          // Skip this property
+          continue;
+        }
+        if (Property.value().is_string()) {
+          auto Value = Property.value().get<std::string>();
+          LOG(6, "kafka broker config {}: {}", Key, Value);
+          Settings.BrokerSettings.ConfigurationStrings[Key] = Value;
+        } else if (Property.value().is_number()) {
+          auto Value = Property.value().get<int64_t>();
+          LOG(6, "kafka broker config {}: {}", Key, Value);
+          Settings.BrokerSettings.ConfigurationIntegers[Key] = Value;
+        } else {
+          LOG(3, "can not understand option: {}", Key);
+        }
+      }
+    }
+  }
+}
+
+void ConfigParser::extractStreamSettings(ConfigSettings &Settings) {
+  using nlohmann::json;
+  if (auto StreamsMaybe = find<json>("streams", Json)) {
+    auto Streams = StreamsMaybe.inner();
+    if (Streams.is_array()) {
+      for (auto const &StreamJson : Streams) {
+        StreamSettings Stream;
+        // Find the basic information
+        extractMappingInfo(StreamJson, Stream.Name, Stream.EpicsProtocol);
+
+        // Find the converters, if present
+        if (auto x = find<nlohmann::json>("converter", StreamJson)) {
+          if (x.inner().is_object()) {
+            Stream.Converters.push_back(extractConverterSettings(x.inner()));
+          } else if (x.inner().is_array()) {
+            for (auto const &ConverterSettings : x.inner()) {
+              Stream.Converters.push_back(
+                  extractConverterSettings(ConverterSettings));
+            }
+          }
+        }
+
+        Settings.StreamsInfo.push_back(Stream);
+      }
+    }
+  }
+}
+
+void ConfigParser::extractMappingInfo(nlohmann::json const &Mapping,
+                                      std::string &Channel,
+                                      std::string &Protocol) {
+  if (!Mapping.is_object()) {
+    throw MappingAddException("Given Mapping is not a JSON object");
+  }
+
+  auto ChannelMaybe = find<std::string>("channel", Mapping);
+  if (!ChannelMaybe) {
+    throw MappingAddException("Can not find channel");
+  }
+  Channel = ChannelMaybe.inner();
+
+  auto ChannelProviderTypeMaybe =
+      find<std::string>("channel_provider_type", Mapping);
+  if (ChannelProviderTypeMaybe) {
+    Protocol = ChannelProviderTypeMaybe.inner();
+  } else {
+    // Default is pva
+    Protocol = "pva";
+  }
+}
+
+ConverterSettings
+ConfigParser::extractConverterSettings(nlohmann::json const &Mapping) {
+  ConverterSettings Settings;
+  Settings.Schema = find<std::string>("schema", Mapping).inner();
```

Comment:
should really have error handling like the rest of the `find` checks.

Reply:
👍